plugins/chat/app/services/chat/list_channel_thread_messages.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+module Chat
+  # List messages of a thread before and after a specific target (id, date),
+  # or fetching paginated messages from last read.
+  #
+  # @example
+  #  Chat::ListThreadMessages.call(thread_id: 2, guardian: guardian, **optional_params)
+  #
+  class ListChannelThreadMessages
+    include Service::Base
+
+    # @!method call(guardian:)
+    #   @param [Integer] channel_id
+    #   @param [Guardian] guardian
+    #   @option optional_params [Integer] thread_id
+    #   @option optional_params [Integer] channel_id
+    #   @return [Service::Base::Context]
+
+    contract
+
+    model :thread
+    policy :ensure_thread_enabled
+    policy :can_view_thread
+    step :determine_target_message_id
+    policy :target_message_exists
+    step :fetch_messages
+
+    class Contract
+      attribute :thread_id, :integer
+      validates :thread_id, presence: true
+
+      # If this is not present, then we just fetch messages with page_size
+      # and direction.
+      attribute :target_message_id, :integer # (optional)
```

Comment:
It really doesn't matter since it's only 2 classes using this stuff, but could be nice to have a way to share these attributes between this class and ListChannelMessages. No need to change, just a thought.

Reply:
yes I thought of it, I need to talk of this with @Flink. Some kind of shared contracts, not sure if it's totally worth it though... will it happen that often?

Reply:
Actually, we can already use a custom contract class and use it in a service:
```ruby
contract :my_contract, class_name: MyContractClass
```

Reply:
We can use multiple contracts right?

Reply:
> We can use multiple contracts right?

Yes, they just have to be named differently.

Reply:
Oh yes I forgot about the contract classes, nice!